File: libs/workflow-core/src/graph/loop-guard.ts
export class LoopGuard {
  private counts = new Map<string, number>();

  constructor(private globalLimit: number) {}

  check(from: string, to: string, edgeLimit?: number): boolean {
    const key = `${from}->${to}`;
    const current = (this.counts.get(key) ?? 0) + 1;
    this.counts.set(key, current);
    const limit = edgeLimit ?? this.globalLimit;
    return current <= limit;
  }

  getLoopMap(): Record<string, number> {
    const map: Record<string, number> = {};
    for (const [key, count] of this.counts) {
      map[key] = count;
    }
    return map;
  }
}

File: apps/controlplane/internal/graph/loopguard.go
package graph

import "fmt"

type LoopGuard struct {
	globalLimit int
	counts      map[string]int64
}

func NewLoopGuard(globalLimit int) *LoopGuard {
	return &LoopGuard{globalLimit: globalLimit, counts: make(map[string]int64)}
}

// Check increments count for from->to, returns true if within limit. edgeLimit 0 = use global.
func (lg *LoopGuard) Check(from, to string, edgeLimit int) bool {
	key := fmt.Sprintf("%s->%s", from, to)
	lg.counts[key]++
	limit := lg.globalLimit
	if edgeLimit > 0 {
		limit = edgeLimit
	}
	return lg.counts[key] <= int64(limit)
}

func (lg *LoopGuard) GetLoopMap() map[string]int64 {
	m := make(map[string]int64, len(lg.counts))
	for k, v := range lg.counts {
		m[k] = v
	}
	return m
}
